datalad_next/tree.py

```diff
@@ -0,0 +1,1130 @@
+
+        # set callable to exclude nodes from the tree, meaning they
+        # will not be yielded by the node generator
+        self.exclude_node_func = exclude_node_func or self.default_exclude_func
+
+        # keep track of levels where the subtree is exhausted,
+        # i.e. we have reached the last child of the current subtree.
+        # this is needed for the custom results renderer, to display
+        # nodes differently based on their relative position in the tree.
+        self.exhausted_levels = set([])
+
+        # store dict with count of nodes for each node type, similar to the
+        # tree command's 'report line' at the end of the output.
+        # the node types (subclasses of ``_TreeNode``) are mutually exclusive,
+        # so the sum of their counts equals to the total node count.
+        # does not count the root itself, only the contents below the root.
+        self.node_count = {node_type.__name__: 0
+                           for node_type in _TreeNode.__subclasses__()}
+
+    @staticmethod
+    def default_exclude_func(path: Path):
+        """By default, exclude files and hidden directories from the tree"""
+        return any((not path.is_dir(), path.name.startswith(".")))
+
+    def path_depth(self, path: Path):
+        """Calculate directory depth of a given path relative to the root of
+        the tree.
+
+        Can also be a negative integer if the path is a parent of the
+        tree root.
+
+        Parameters
+        ----------
+        path: Path
+
+        Returns
+        -------
+        int
+            Number of levels of the given path *below* the tree root (positive
+            integer) or *above* the tree root (negative integer)
+        """
+        if is_path_relative_to(path, self.root):
+            return len(path.relative_to(self.root).parts)
+        elif is_path_relative_to(self.root, path):
+            return - len(self.root.relative_to(path).parts)
+        else:
+            raise ValueError("Could not calculate directory depth: "
+                             f"'{path}' is not relative to the tree root "
+                             f"'{self.root}' (or vice-versa)")
+
+    def is_recursive_symlink(self, dir_path: Path):
+        """Detect symlink pointing to a directory within the same tree
+        (directly or indirectly).
+
+        The default behaviour is to follow symlinks. However, we do not follow
+        symlinks to directories that we may visit or have visited already,
+        i.e. are also located under the tree root or any parent of
+        the tree root (within a distance of ``max_depth``).
+
+        Otherwise, the same subtree could be generated multiple times in
+        different places, potentially in a recursive loop (e.g. if the
+        symlink points to its parent).
+
+        This is similar to the logic of the UNIX 'tree' command, but goes a
+        step further to prune all duplicate subtrees.
+        """
```

Comment:
When trying out failure scenarios (such as your `[Errno 40] Too many levels of symbolic links` example), I came across the case of a symlink pointing to a directory within the tree. This caused the recursive generation of the same subtree for 33 (!) times (I think that's the value of `MAXSYMLINKS` [on MacOS](https://github.com/apple-oss-distributions/xnu/blob/xnu-4903.270.47/bsd/sys/param.h#L207)). While it may be an edge case, `tree` can handle it, so we should too. The goal is to not generate any duplicate nodes.

See tests: https://github.com/datalad/datalad-next/blob/de16e02068ce565cfbf697ca86d1e7ce95233eef/datalad_next/tests/test_tree.py#L753-L763

Reply:
Strictly speaking, a symlink is a file (even when it points to a directory). In terms of datalad result records, I'd give such a node a `type="symlink"` (neither file nor directory). In terms of counting for the sake of summary rendering I have no preference, as long as the structured records are precise. For that, it may make sense to record the fact that a symlink is pointing to a directory. I have currently no splendid idea how that should be done. I will keep thinking and return here.

Reply:
For the purpose of tree display, I think the link's target should dictate the node type and summary count. While the symlink itself is always technically a file, we follow symlinks when traversing the tree, so it makes sense to count links pointing to directories as directories (`tree` does this too, even without the `-l` option to follow links).

However for the results record, you are right that we may want the node type (unless it's a dataset) to match the actual file type -- what you'd get from `ls` or `stat -f "%T"` in Linux. Then for a symlink pointing to a directory we'd have `type=symlink, is_directory=True`. This is not too different from the current `type=directory, is_symlink=True`... but has the advantage of aligning with the file type model, so could make it more intuitive to use for consumers.